File: Data Structures & Algorithms/sliding-window-maximum/submission-4.cpp
class Solution {
public: 
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> result;
        result.reserve(nums.size() - k + 1);

        deque<int> indexDeque; // Sorted decreasing value from index
        for(int currentIndex = 0; currentIndex < nums.size(); currentIndex++){
            int currentNumber = nums[currentIndex];

            while(!indexDeque.empty() && indexDeque.front() <= currentIndex - k){
                indexDeque.pop_front();
            }

            while(!indexDeque.empty() && currentNumber > nums[indexDeque.back()]){
                indexDeque.pop_back();
            }
            
            indexDeque.emplace_back(currentIndex);

            if(currentIndex >= k - 1){
                result.emplace_back(nums[indexDeque.front()]);
            }
            
        }

        return result;
    }
};

File: Data Structures & Algorithms/string-matching-in-an-array/submission-4.py
class Solution:
    def stringMatching(self, words: List[str]) -> List[str]:
        result: List[str] = []
        words = sorted(words, key=len)
        for index in range(len(words)):
            for innerIndex in range(index + 1, len(words)):
                if words[innerIndex].find(words[index]) != -1:
                    result.append(words[index])
                    break

        return result